prometheusbp/high_watermark_gauge.go

```diff
@@ -0,0 +1,58 @@
+package prometheusbp
+
+import (
+	"sync"
+)
+
+// HighWatermarkGauge implements an int64 gauge with high watermark value.
+type HighWatermarkGauge struct {
+	lock sync.RWMutex
```

Comment:
benchmark result with `RWMutex`:
```
goos: linux
goarch: amd64
pkg: github.com/reddit/baseplate.go/prometheusbp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkHighWatermarkGauge-12          20724597                55.75 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/reddit/baseplate.go/prometheusbp     1.221s
```

benchmark result if we use `Mutex` instead:
```
goos: linux
goarch: amd64
pkg: github.com/reddit/baseplate.go/prometheusbp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkHighWatermarkGauge-12          12081015               103.0 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/reddit/baseplate.go/prometheusbp     1.351s
```